src/Magento/FunctionalTestingFramework/Extension/PageReadinessExtension.php

```diff
@@ -0,0 +1,227 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+namespace Magento\FunctionalTestingFramework\Extension;
+
+use Codeception\Event\StepEvent;
+use Codeception\Event\TestEvent;
+use Codeception\Events;
+use Codeception\Exception\ModuleRequireException;
+use Codeception\Extension;
+use Codeception\Module\WebDriver;
+use Codeception\TestInterface;
+use Facebook\WebDriver\Exception\UnexpectedAlertOpenException;
+use Magento\FunctionalTestingFramework\Extension\ReadinessMetrics\AbstractMetricCheck;
+use Facebook\WebDriver\Exception\TimeOutException;
+use Magento\FunctionalTestingFramework\Util\Logger\LoggingUtil;
+use Magento\FunctionalTestingFramework\Config\MftfApplicationConfig;
+use Monolog\Logger;
+
+/**
+ * Class PageReadinessExtension
+ */
+class PageReadinessExtension extends Extension
+{
+    /**
+     * Codeception Events Mapping to methods
+     *
+     * @var array
+     */
+    public static $events = [
+        Events::TEST_BEFORE => 'beforeTest',
+        Events::STEP_BEFORE => 'beforeStep',
+        Events::STEP_AFTER => 'afterStep'
+    ];
+
+    /**
+     * @var Logger
+     */
+    private $logger;
+
+    /**
+     * Logger verbosity
+     *
+     * @var bool
+     */
+    private $verbose;
+
+    /**
+     * Array of readiness metrics, initialized during beforeTest event
+     *
+     * @var AbstractMetricCheck[]
+     */
+    private $readinessMetrics;
+
+    /**
+     * Active test object
+     *
+     * @var TestInterface
+     */
+    private $test;
```

Comment:
That depends on how Codeception handles events.  I use the test object from the beforeTest event to track metric failure data ( $test->getMetadata()->setCurrent() ), so if the test object attached to the step events gets recreated or the metadata isn't otherwise persisted between events then that tracking functionality is lost.

I could move that tracking out of the test object and into an extension-specific construct directly instead, though I'd still be passing that around between metrics.  Any objections?

Reply:
Hmm that's a good point. I guess I'd only be concerned about tethering ourselves to codeception's test object for your metadata, but I don't feel that strongly about it atm. @okolesnyk or @magterskine any thoughts?

Reply:
@pdohogne-magento @imeron2433 @magterskine 
I think in our case when we need to release this next week we can leave it as it is for now. And probably do the refactoring when the story about skip metrics attribute for actions will be in progress.